### PR TITLE
Issue #194 added method for engine warm up and test

### PR DESCRIPTION
**Pull request by theaaron123**
#194

### CODE REVIEW

**Comment by karussell** [comment]
Thanks!  As a start this is good to go. Technically I would just have it into a separate class to avoid growing GraphHopper.java even more. And also removing the test ... as this is not really a test and not really testable I fear.

And also idea is simple&good but would fail for certain areas. So as a second step we need something more graph based (e.g. a BFS per subnetwork) that warms up the whole graph. Surely this would need special care as it would be forbidden for non-CH requests. Maybe then we could use BFS and skip certain nodes randomly then.

**Comment by karussell** [comment]
Thanks, this looks nice! 

Sorry for my trouble and my incompleteness at the first comment: would you mind to do three more changes (should be the last :) !!):

 * make it possible via CH with your old query method but instead of lat,lon pick random node ids and catch exceptions and retry if e.g. different subnetworks (I think a warmup should never fail with an exception?)
 * limit BFS search to max_visited_nodes

File: core/src/main/java/com/graphhopper/util/EngineWarmUp.java [hunk 1]
```diff
@@ -1,26 +1,51 @@
...
+    }
+
+    private static void warmUpNonCHSubNetwork(GraphHopper graphHopper, int iterations) {
+        GraphHopperStorage ghStorage = graphHopper.getGraphHopperStorage();
+
+        BreadthFirstSearch bfs = new BreadthFirstSearch();
         for (int i = 0; i < iterations; i++) {
-            int startNode = rand.nextInt(ghStorage.getBaseGraph().getNodes() + 1);
+            int startNode = ThreadLocalRandom.current().nextInt(0, graphHopper.getMaxVisitedNodes() + 1);
             bfs.start(ghStorage.getBaseGraph().createEdgeExplorer(), startNode);
```
Comment (karussell) [inline core/src/main/java/com/graphhopper/util/EngineWarmUp.java]: The previous nextInt code is correct. The BFS needs to stop after graphHopper.getMaxVisitedNodes() has reached i.e. something like:

```
// it is save to create this once and reuse for others as we have no different states or threads involved:
EdgeExplorer explorer = ghStorage.getBaseGraph().createEdgeExplorer()
for loop {
  // fresh instance necessary to reset counter every time
  new BreadthFirstSearch() {
     int counter = 0;
     public boolean goFurther(int nodeId) {
        counter++;
        return counter < maxVisitedNodes;
     }
  }.start(edgeExplorer, startNode)
}
```

This is not entirely correct as we can visit certain nodes multiple times but we would have to call `bitset.getCardinality` on every call which is too expensive and so not necessary for this 'warmup' purpose.

**Comment by karussell** [comment]
Thanks - would you mind to add catching all exceptions?

**Comment by karussell** [comment]
Didn't mean to rethrow ... the problem could be that you pick one point from the mainland and one from the island with no connection. So just catch the exception and log it. But no counter & no rethrow necessary - I also do not know of any example where a 2nd request should succeed if the previous one failed.

**Comment by theaaron123** [comment]
@karussell  sorry, log by printing the stack trace, creating an instance of logger, or by throwing with an error message?

**Comment by karussell** [comment]
creating an instance of Logger via slf4j LoggerFactory

**Pull request closed without merging**